### ISSUE TITLE
Typescript cannot detect mismatching between two type declaration of same t function

### ISSUE THREAD

**Issue by skmohammadi** [i18next/react-i18next#1502]
## 🐛 Bug Report

Typescript cannot detect mismatching between two type declaration of same **t** function

## To Reproduce

Consider a function with the following definition:

```js
const myFunction = (t: TFunction<'nsB'[]>) => {
/// code
}
```

Take a look at the following code:

```js
const { t } = useTranslation(['nsA']);
myFunction(t)
```

Why typescript not give any error?

## Expected behavior

The following error should occur from TS:

```js
Argument of type "nsA"[] is not assignable to parameter of type "nsB"[].
```

## Your Environment

- *runtime version*: i.e. node v16
- *i18next version*: 11.0.0
- *Framework*: Next 12
- *os*: Windows

**Comment by skmohammadi** [comment]
@pedrodurek thank you. So is there any solution or trick?

**Comment by skmohammadi** [comment]
I found a tricky solution, but its need to override **TFunction** interface. How can I do it?

**Comment by skmohammadi** [comment]
@pedrodurek The tricky solution is overriding **TFunction** interface as below:

```ts
export interface TFunction<N extends Namespace = DefaultNamespace, TKPrefix = undefined, F = N> {
  <
    TKeys extends TFuncKey<N, TKPrefix>,
    TDefaultResult extends TFunctionResult | React.ReactNode = string,
    TInterpolationMap extends object = StringMap,
  >(
    key: TKeys | TKeys[],
    options?: TOptions<TInterpolationMap> | string,
    ns?: F,
  ): TFuncReturn<N, TKeys, TDefaultResult, TKPrefix>;
  <
    TKeys extends TFuncKey<N, TKPrefix>,
    TDefaultResult extends TFunctionResult | React.ReactNode = string,
    TInterpolationMap extends object = StringMap,
  >(
    key: TKeys | TKeys[],
    defaultValue?: string,
    options?: TOptions<TInterpolationMap> | string,
    ns?: F,
  ): TFuncReturn<N, TKeys, TDefaultResult, TKPrefix>;
}
```

New generic argument **F** added

**Comment by pedrodurek** [comment]
Hey @skmohammadi, that's not an ideal solution because we'd be adding an inexistent argument to `T` function. I'll explore other options and let you know.

**Comment by skmohammadi** [comment]
@pedrodurek ok but is there any standard solution?

**Comment by yekver** [comment]
This issue is still reproducible, are there any generic solution to this problem?

**Comment by anhnch** [comment]
Since version 12 react-i18next use TFunction from i18next and discarded the fix in v11.16.4
```
TDefaultResult extends TFunctionResult | React.ReactNode = string,
```
This comes back
<img width="381" alt="image" src="https://user-images.githubusercontent.com/131125/209753179-732bfac7-0562-44fc-a627-120c1940dc34.png">
I patch the i18next temporarily, but I hope someone can fix this because React.ReactNode should be on the react-i18next side
